units/patch_intent.go

```diff
@@ -527,30 +525,38 @@ func (j *patchIntentProcessor) setToPreviousPatchDefinition(patchDoc *patch.Patc
 		if err = setTasksToPreviousFailed(patchDoc, previousPatch, project); err != nil {
 			return "", errors.Wrap(err, "settings tasks to previous failed")
 		}
-
 	} else {
 		patchDoc.Tasks = previousPatch.Tasks
 	}
 
 	return previousPatch.Status, nil
 }
 
-func getPreviousFailedTasksAndDisplayTasks(tasksInProjectVariant []string, vt patch.VariantTasks, version string) ([]string, error) {
+func getPreviousFailedTasksAndDisplayTasks(project *model.Project, vt patch.VariantTasks, version string) ([]string, error) {
+	tasksInProjectVariant := project.FindTasksForVariant(vt.Variant)
 	failedTasks, err := task.FindAll(db.Query(task.FailedTasksByVersionAndBV(version, vt.Variant)))
 	if err != nil {
 		return nil, errors.Wrapf(err, "finding failed tasks in build variant '%s' from previous patch '%s'", vt.Variant, version)
 	}
-	failedExecutionTasks := []string{}
+	// Verify that the task group or task is in the current project definition and in the previous run.
+	allFailedTasks := []string{}
 	for _, failedTask := range failedTasks {
-		if !failedTask.DisplayOnly {
-			failedExecutionTasks = append(failedExecutionTasks, failedTask.DisplayName)
+		if utility.StringSliceContains(vt.Tasks, failedTask.DisplayName) {
+			if failedTask.TaskGroup != "" &&
+				utility.StringSliceContains(tasksInProjectVariant, failedTask.TaskGroup) {
+				if failedTask.IsPartOfSingleHostTaskGroup() {
```

Comment:
Can we add a comment to this to indicate why we're adding earlier task group tasks in the single host case?

Reply:
In general since these clauses are long it would be helpful to have a comment for each of them; comments are free!